src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs

```diff
@@ -219,45 +219,72 @@ private SafeProcessHandle GetProcessHandle()
             {
-                filename = ResolvePath(startInfo.FileName);
                 argv = ParseArgv(startInfo);
+
+                // Invoke the shim fork/execve routine.  It will create pipes for all requested
+                // redirects, fork a child process, map the pipe ends onto the appropriate stdin/stdout/stderr
+                // descriptors, and execve to execute the requested process.  The shim implementation
+                // is used to fork/execve as executing managed code in a forked process is not safe (only
+                // the calling thread will transfer, thread IDs aren't stable across the fork, etc.)
+                result = Interop.Sys.ForkAndExecProcess(
+                    filename, argv, envp, cwd,
+                    startInfo.RedirectStandardInput, startInfo.RedirectStandardOutput, startInfo.RedirectStandardError,
+                    out childPid,
+                    out stdinFd, out stdoutFd, out stderrFd, shouldThrow: !startInfo.UseShellExecute);
             }
 
-            string[] envp = CreateEnvp(startInfo);
-            string cwd = !string.IsNullOrWhiteSpace(startInfo.WorkingDirectory) ? startInfo.WorkingDirectory : null;
+            if (result != 0)
+            {
+                if (!startInfo.UseShellExecute)
+                {
+                    // Could not find the file
+                    Debug.Assert(string.IsNullOrEmpty(filename));
+                    throw new Win32Exception(Interop.Error.ENOENT.Info().RawErrno);
+                }
+
+                // this time, set the filename as default program to open file/url
+                filename = GetPathToOpenFile();
+                argv = ParseArgv(startInfo, GetPathToOpenFile());
 
-            // Invoke the shim fork/execve routine.  It will create pipes for all requested
-            // redirects, fork a child process, map the pipe ends onto the appropriate stdin/stdout/stderr
-            // descriptors, and execve to execute the requested process.  The shim implementation
-            // is used to fork/execve as executing managed code in a forked process is not safe (only
-            // the calling thread will transfer, thread IDs aren't stable across the fork, etc.)
-            int childPid, stdinFd, stdoutFd, stderrFd;
-            Interop.Sys.ForkAndExecProcess(
-                filename, argv, envp, cwd,
-                startInfo.RedirectStandardInput, startInfo.RedirectStandardOutput, startInfo.RedirectStandardError,
-                out childPid,
-                out stdinFd, out stdoutFd, out stderrFd);
+                result = Interop.Sys.ForkAndExecProcess(
+                    filename, argv, envp, cwd,
+                    startInfo.RedirectStandardInput, startInfo.RedirectStandardOutput, startInfo.RedirectStandardError,
+                    out childPid,
+                    out stdinFd, out stdoutFd, out stderrFd);
+
+                Debug.Assert(result == 0);
```

Comment:
If result != 0 does that mean you made a mistake in your code? If not, Debug.Assert is probably not the right choice. You should either ignore it (if appropriate) or throw, or some other effect visible in retail.

Reply:
it is not expected to reach here and result != 0 so I think the assert is proper here.

Reply:
ok fine